odt2txt: Convert openDocuments to txt

albfan opened the pull request:
Add odt2txt to use it on git-for-windows

@dscho suggest this is better place for https://github.com/Alexpux/MSYS2-packages/pull/964

albfan commented:
@dscho https://tea-ci.org/repos/Alexpux/MINGW-packages/builds/3462 fails. That means that regex dep belongs to a POSIX environment?

dscho commented:
> That means that regex dep belongs to a POSIX environment?

It kinda does. But we already use PCRE in Git for Windows, and PCRE ships with that nice POSIX compatibility layer. You'll just need to make sure that `pcreposix.h` is included, and that it is linked with `-lpcreposix`.

albfan commented:
Then I will close this PR. In msys packages it don't need that link. Why?

elieux commented:
Wait, what about `mingw-w64-pcre` and `mingw-w64-pcre2`, are they not sufficient?

elieux commented:
I think *that*'s what Johannes was trying to point to.

albfan commented:
Oh, I reread several times its comment, but can find the solution. Let me try again

albfan commented:
Ok that seems to work @elieux 

@dscho @Alexpux  Is this them the correct place for odt2txt?

albfan commented:
What? https://ci.appveyor.com/project/Alexpux/mingw-packages/build/1.0.3424#L486

elieux commented:
Something weird with the CI.

elieux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,34 @@
+arch=('any')
+depends=('zlib-devel' 'libiconv-devel' "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+makedepends=('git')
+source=("$pkgname::git://github.com/dstosberg/odt2txt.git#tag=v${pkgver}" "0001-Compile-for-Msys2.patch")
```
Why complicate it with Git? How about `https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz`?

albfan replied:
done

elieux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,34 @@
+
+pkgname=odt2txt
+pkgver=0.5
+pkgrel=2
```
Set `pkgrel` to `1` for brand new packages.

albfan replied:
done

elieux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,34 @@
+[[ "$CARCH" == "x86_64" ]] && arch_make=64
+[[ "$CARCH" == "i686" ]] && arch_make=32
+
+pkgname=odt2txt
```
Alter the name according to conventions for mingw-packages,

albfan replied:
done

elieux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,34 @@
+# Maintainer: Alberto Fanjul <albertofanjul@gmail.com>
+
+[[ "$CARCH" == "x86_64" ]] && arch_make=64
+[[ "$CARCH" == "i686" ]] && arch_make=32
```
Probably move these two lines into `build()` where it's used.

albfan replied:
done

elieux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,34 @@
+  mkdir -p "$pkgdir"/usr/bin
+  cp odt2txt.exe "$pkgdir"/usr/bin/odt2txt.exe
+  mkdir -p "$pkgdir"/usr/share/man/man1
+  cp odt2txt.1 "$pkgdir"/usr/share/man/man1/odt2txt.1
```
Use correct variables to install into the right `/mingw*` directory, see any other `PKGBUILD`.

albfan replied:
check mingw-w64-ag, mingw-w64-docbook-sgml and several others and don't have idea of what to do.

elieux replied:
Copy stuff into `${pkgdir}${MINGW_PREFIX}/` instead of `${pkgdir}/usr/`.

dscho commented:
> What? https://ci.appveyor.com/project/Alexpux/mingw-packages/build/1.0.3424#L486

It just means that the CI has been set up incorrectly:

> ==> ERROR: A package has already been built. (use -f to overwrite)

This means that the package was already built. If CI is set up correctly, there is no file in the source directory apart from the tracked files and the `.git/` directory. Apparently previous checkouts are re-used, *without* calling `git clean` first. Not your problem, @albfan

elieux commented:
Yeah. I didn't mean that the CI platform was broken, but the setup was weird in general.

albfan commented:
Cool, cool, cool. Let me rework those reviews, seems we are almost there.

albfan commented:
@dscho I guess the error is related with a .gitignore I added. Maybe it is using git clean -df, but it needs git clean -dxf.

Going to upload what I have by now, and see. There's a thing about install i don't understand how to fix

albfan commented:
Ok, now I can reproduce locally. I added:
 
    $ export MINGW_PACKAGE_PREFIX=mingw-w64-x86_64
    $ export PKG_CONFIG_PATH=/mingw64/lib/pkgconfig

without luck, but supposed on CI it will work (it does!)

@elieux I still didn't fix the install part.

elieux commented:
@albfan, are you using `makepkg-mingw` locally to build the package?

albfan commented:
Tip saved. Didn't know about that.

What about the package part? Didn't understand what I need to fix

Alexpux commented:
@albfan you must not use ````/usr```` prefix for mingw packages. Also download source archive into named tarball

Alexpux commented on mingw-w64-odt2txt/0001-Compile-for-Msys2.patch:
```diff
@@ -0,0 +1,56 @@
+ 	LIBS += -liconv
+ 	EXT = .exe
+ endif
++ifeq ($(UNAME_O),Msys)
```
I think this is wrong. When building mingw packages UNAME will return MINGW32 or MINGW64

albfan replied:
Oh yes that was when this PR was for msys repo. Have to try on a minute environment

Alexpux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,32 @@
+url="https://github.com/dstosberg/odt2txt/"
+license=("GPL2")
+arch=('any')
+depends=('zlib-devel' 'libiconv-devel' "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
```
Mingw packages can't depends on msys packages

albfan replied:
I will try to find the replacement

Alexpux commented:
I'm add more comments to fix

Alexpux commented on mingw-w64-odt2txt/0001-Compile-for-Msys2.patch:
```diff
@@ -0,0 +1,56 @@
+ endif
++ifeq ($(UNAME_O),Msys)
++	CFLAGS += -I/mingw$(ARCH)/lib/libzip/include
++	LIBS += -liconv -lpcreposix -lpcre -llibzip -lzip -lz -L/mingw$(ARCH)/lib
```
You don't need to specify directory for libraries search as GCC know it.

albfan replied:
I guess is another flaw to not use makepkg-mingw. Will fix it

albfan commented:
I fixed package step. Removed man install as seems is not helpful on windows.

Tea CI has jobs pending from 15 hours ago, so will take a while to see it passing

Don't know how to make it "download into named tarball", examples?

Alexpux commented:
See for example:
https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-lua-mpack/PKGBUILD#L17

albfan commented:
Crossing fingers

Alexpux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,32 @@
+arch=('any')
+depends=("mingw-w64-${CARCH}-libiconv" "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+source=("${_realname}-${pkgver}.tar.gz"::"https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz" "0001-Compile-for-Msys2.patch")
+sha256sums=('23a889109ca9087a719c638758f14cc3b867a5dcf30a6c90bf6a0985073556dd' 'SKIP') 
```
One checksum per line.
Use ````updpkgsums```` to recalculate chechsums

albfan replied:
Sure. Afk now, that's why I skipped by now

Alexpux commented on mingw-w64-odt2txt/PKGBUILD:
```diff
@@ -0,0 +1,32 @@
+license=("GPL2")
+arch=('any')
+depends=("mingw-w64-${CARCH}-libiconv" "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+source=("${_realname}-${pkgver}.tar.gz"::"https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz" "0001-Compile-for-Msys2.patch")
```
Use ````pkgver```` variable instead hardcoded version

Alexpux replied:
One source per line

albfan replied:
Cool

Alexpux commented:
Update checksums and fix patch for mingw

Alexpux commented:
Will you fix not resolved issues?

albfan commented:
I will try an online service to find the sha256 for patch

Alexpux commented:
You need just run updpkgsums inside folder with PKGBUILD to generate new checksums.
Also you need fix patch as UNAME don't return Msys when build mingw packages

albfan commented:
@Alexpux I'm afk. Will took me some commits and looks at CI output to fix all that. If you can do it I don't mind. All this should besquashed in a single commit when merging

albfan commented:
I failed to find sha256 for patch (I guess is the typical \r\n problem when copy paste) cannot fix that even from mobile!

Alexpux commented:
@albfan you dont need to find sha256 utility - use ````updpkgsums```` command

albfan commented:
I know that, but I have no laptop to do that, sorry!

albfan commented:
Great!. Commit message is total crap! @dscho hope we can continue from here with fit for windows

@alexpux many thanks for all the support. Hope future PR will be straight merged

dscho commented:
> Update checksums and fix patch for mingw

@Alexpux please note that your terse prose is prone to be misunderstood. Probably not what you want. The solution would be to be precise enough to make yourself clear.

In this case, it would have saved a useless round-trip and a few hours to specify the commands you expect to work.

In this case, what you *should have wanted* to say is:

1. Please run `updpkgsums` in the same directory as `PKGBUILD` to update the checksums.
2. I still believe that `uname -o` outputs `MINGW32` in MINGW mode (which is the mode in which `MINGW-packages` is supposed to be built) and hence the `.patch` file should be adjusted accordingly.

It can't be fun for you to waste so much of your own time simply by being so unclear as to not get your points across.

Also, I cared to test your claim about "UNAME" (which really was `uname -o`, I guess) and it is incorrect:

```sh
$ uname -o
MSys
```

To make sure, this is with a regular `msys2_shell.cmd`, i.e. *not* `msys2_shell.cmd -msys`. So the original patch was correct and your suggestion was wrong, but the review was also confusing so I do not fault @albfan for being unable to make sense of it.

Alexpux commented:
@dscho makepkg-mingw change output of uname -o

dscho commented:
@Alexpux oh wow. How? And why have you not mentioned this earlier??? This is another completely unnecessary round-trip. Just like the next unnecessary round-trip: after merging this, what is you time-line to release this to the Pacman repository? I feel you think that this short-hand type of conversation saves you time. I disagree. It wastes your time as much as it wastes mine. And every other contributor's.

Alexpux commented:
@dscho environment variable ````MSYSTEM```` is responsible for output of ````uname -o````. This is true for MSYS1 too.

dscho commented:
I just updated, just to make sure. and I made sure to start with `msys2_shell.cmd -mingw64` explicitly.

```sh
me@work MINGW64 ~
$ echo $MSYSTEM
MINGW64

me@work MINGW64 ~
$ uname -o
Msys
```

Yet another round-trip. This is becoming beyond annoying. What a deliberate waste of my time!

Alexpux commented:
You are right, MSYSTEM change ````uname -s```` output.

albfan commented:
I'm maintainer for other open source projects. I see all these questions and answers as knowledge sharing. I know few to none about MinGW and Msys, but all these info could help to fill the gap between my *nix user experience and my windows user experience, which is still really poor.

Hope to contribute with more packages, so understand all the workflow would be nice. (Pacman repo, etc) if there's something I can help with I'm on IRC too

dscho commented:
@albfan it is really good to hear that you do not give up so easily. And I really hope that @Alexpux will eventually see the merit of my suggestions to improve his communication style.

mingwandroid commented:
@dscho, not every speaks fluent English. I do not see anyone having a communication style problem here except yourself. Did you get out of the wrong side of the bed again today?

People contribute to Open Source because it is their passion, abuse such as you give out is a sure way to throw water on that passion.

mingwandroid commented:
Further I would expect people to do some research before setting out to build packages with a system they are not familiar with. That research would, in the case of MSYS2, involve reading https://github.com/msys2/msys2/wiki/Creating-packages

It's not a problem when people do not find this page, we'll still try to help them out.

Alexpux commented:
@albfan @dscho I think better to do in patch somethink like this
````ifneq (,$(findstring MINGW,$(UNAME_S)))````
instead
````ifeq ($(UNAME_O),Msys)````

albfan commented:
@Alexpux muchas better that way. Feels weird that comparing to "Msys" works for a MinGW package

@mingwandroid thanks for sharing. I was sure there was docs about setting up MinGW. I use to work that way (read first push when is done) Is just windows adds another layer of complexity I can deal with (it gets better with practice).

If you can wait for this I can rework it as it should be. I'm not able to do it right now

Thanks to all for your patience. I'm really enjoying this master class

Alexpux commented:
@albfan I'm fix mingw detection and installation in https://github.com/Alexpux/MINGW-packages/commit/f66df107d73edcd10659297e98c5566572259110

albfan commented:
Terrific!. Seems easy when an expert get it done.

dscho commented:
> I'm fix mingw detection and installation in f66df10

Perfect. This message was clear. As to the fix: it is identical to how Git itself does it: https://github.com/git-for-windows/git/blob/v2.14.1.windows.1/config.mak.uname#L594, so I am convinced it is correct.
